pkg/nsx/services/compare.go

```diff
@@ -49,22 +49,57 @@ func RulesEqual(existingRules []model.Rule, rules []model.Rule) (bool, []model.R
 func RulesEqualDetail(existingRules []model.Rule, rules []model.Rule) bool {
 	isEqual := true
 	for i := 0; i < len(rules); i++ {
-		r1, _ := simplifyRule(&existingRules[i]).GetDataValue__()
-		r2, _ := simplifyRule(&rules[i]).GetDataValue__()
-		var dataValueToJSONEncoder = cleanjson.NewDataValueToJsonEncoder()
-		s1, _ := dataValueToJSONEncoder.Encode(r1)
-		s2, _ := dataValueToJSONEncoder.Encode(r2)
-		if s1 != s2 {
-			log.Info("", "nsx rule", s1, "k8s rule", s2)
+		// if we use dataValueToJSONEncoder to marshal the whole rule, it will lose the "ServiceEntries" property,
+		// it shows that the result is "service_entries":[]" or even nothing, so everytime it shows not equal and
+		// operate nsx-t repeatedly, perhaps the sdk failed to marshal the nested property which is a list of
+		// *data.StructValue. We suppose *data.ListValue is the right way to do it.
+		// however, the ServiceEntries which model.Rule contains is []*data.StructValue, rather than *data.ListValue,
+		// we don't know why sdk design like this, We leave it for follow-up observation.
+		// now we fix this problem by comparing them separately.
+		r1, _ := json.Marshal(simplifyRule(&existingRules[i]))
+		r2, _ := json.Marshal(simplifyRule(&rules[i]))
+		if string(r1) != string(r2) {
+			log.Info(
+				"compare rule",
+				"nsx rule",
+				simplifyRule(&existingRules[i]),
+				"k8s rule",
+				simplifyRule(&rules[i]),
+			)
 			isEqual = false
 			break
 		}
+
+		se1 := existingRules[i].ServiceEntries
+		se2 := rules[i].ServiceEntries
+		if len(se1) != len(se2) {
+			log.Info(
```

Comment:
I don't think the way to print log in so many lines, is it golint style?

Reply:
golines style.